### DISCUSSION TITLE
automatic collection name mapping

### DISCUSSION THREAD

**Discussion by pavlis** [mspass-team/mspass#144] [Design & Development]
I ran into a collision I'm not sure how to handle.  The Database.read_data function uses the normalized attribute feature in the Schema and maps attributes from the channel collection into names like:  channel_lat, channel_lon, etc.   Those are good because otherwise lat and lon would be ambiguous.  The problem I just encountered is "hang" and "vang".  They get mapped to "channel_hang" and "channel_vang" respectively.  That change is breaking the C++ constructor called in the bundle algorithm.  It has hard wired Metadata get method calls asking for "hang" and "vang".   I could fix this by changing the C++ code, but it brings up a general problem. This is a type example of where an alias is needed to match internal requirements of a piece of code you don't want to mess with or maybe you cannot at all as the code is proprietary (e.g. reflection processing systems).  

What is the recommended way to handle this?   Metadata has a "change_key" method which is the efficient way to handle it, but seems we should have a standard python function to apply a set of registered aliases for a particular use.  This one is pretty trivial, but there are nontrival examples like a SAC alias definition that would do the mapping with a single call.   Wonder if we should plan an aliases module to handle some of this kind of thing?   A generic method for a simple mapping between namespaces could have this signature:
```
def apply_aliases(md, aliases={"channel_hang" : "yang, "channel_vang" : "vang"}):
```
where I made the default here what I'm discussing just for an example not as a suggested default.  Point is the aliases dict could be used in a simple loop to change all keys with a loop like this:
```
# mspass_object is any mspass_object with Metadata as a base class
for key in aliases:
  if key in mspass_object:
    mspass_object.change_key(key,aliases[k])
```

Is there a way to use decorators to define some standard aliases that would run the function using a fixed set of names known to match the schema?

**Comment by wangyinz** [comment]
This is already implemented: https://github.com/wangyinz/mspass/blob/5692907043fbe5f43b06488f80f268b6b3224ca6/python/mspasspy/db/schema.py#L109

**Reply by wangyinz** [reply]
This discussion is duplicated with https://github.com/wangyinz/mspass/discussions/133

**Reply by pavlis** [reply]
Hmm - not sure how I missed this.   Shows why even for internal use MsPASS is getting a large enough code base we need to really get some key missing documentation written.  This one should be discussed in the Database and Schema document for the User's Manual.

**Comment by wangyinz** [comment]
> Is there a way to use decorators to define some standard aliases that would run the function using a fixed set of names known to match the schema?

Yes, we can certainly add that.

btw, let's change our C++ code to use the new attribute naming convention where all the normalized attributes are named as "collection_key".

**Reply by pavlis** [reply]
The problem with hang and vang may be just the tip of the iceberg on this problem.   I think I will do a scan of the code and consider building an include file containing mspass names so usages like that are all in one place.  The file would be a long const string definitions like this:
```
const string SEISMICMD_hang("channel_hang");
const string SEISMICMD_vang("channel_vang");
```
where the point of the "SEISMIC_"  tag is to make unique variable names that would avoid collisions.   The all caps would make them easy to see in the code and might improve maintainability.
